26.6.1 follow-ups: fix `release.yml` artifact download path + add per-release codenames

Two related 26.6.1 items, tracked together. **Part A** is a release-blocking
CI/release bug (must land before the `v26.6.1` tag); **Part B** is an additive
feature (per-release codenames). They are independent , but share this issue.

---

## Part A — `release.yml`: `download-artifact-verified` uses a relative `path: dist` → "List downloaded packages" fails

### Symptom

After #2224 merged to master, the `release` workflow ran on master
(development/nightly build) and failed in the **release-development** job at the
*"List downloaded packages"* step:

```
Downloaded packages:
ls: cannot access 'dist/': No such file or directory
Error: Process completed with exit code 2.
```

(run: https://github.com/crossbario/crossbar/actions/runs/27901609165/job/82562823017)

This is a CI/release bug only (the package itself is fine), but it is
**release-blocking**: the same defect is present in the **release-production**
job, so it would also break the actual PyPI publish triggered by the `v26.6.1`
tag.

### Root cause

`release.yml` passes a **relative** path to the `download-artifact-verified`
shared action:

```yaml
- name: Download distribution artifacts (verified)
  uses: wamp-proto/wamp-cicd/actions/download-artifact-verified@main
  with:
    name: ${{ needs.check-main-workflow.outputs.artifact_dist }}
    path: dist          # <-- relative
    run-id: ${{ needs.check-main-workflow.outputs.main_run_id }}
```

The `download-artifact-verified` action downloads into a staging temp dir,
`cd`s into it to verify checksums, and then (still inside the staging dir) runs
`mkdir -p "$DOWNLOAD_PATH"` + copies the verified files there. Because
`DOWNLOAD_PATH` is the **relative** `dist`, it resolves to `$STAGING_DIR/dist` —
and the action's `trap cleanup_staging EXIT` then deletes the staging dir. So
`$GITHUB_WORKSPACE/dist` is never created.

The download step still exits 0 (it did copy the files — into the doomed staging
path), so `if: success()` lets the next step run, and `ls -lh dist/` (back in
`$GITHUB_WORKSPACE`) fails with *No such file or directory*.

The working reference repos (autobahn-python, wamp-xbr) use an **absolute** path
and do not hit this: `path: ${{ github.workspace }}/dist`.

This was introduced in #2224 when the release jobs were switched from rebuilding
(`just dist`) to downloading the verified artifact from `main.yml`
(chain-of-custody); the absolute-path form from the gold-standard template was
not carried over.

### Fix

Make all `dist` references in `release.yml` absolute, exactly matching
wamp-xbr / autobahn-python. In both the `release-development` and
`release-production` jobs:

- `download-artifact-verified` → `path: ${{ github.workspace }}/dist`
  (this is the critical one)
- `List downloaded packages` → `ls -lh ${{ github.workspace }}/dist/`
- `check-release-fileset` → `distdir: ${{ github.workspace }}/dist`
  (all occurrences, incl. the PyPI "Validate and clean release fileset" step)

(Only the download `path:` is strictly broken — the `ls`/`distdir` steps run in
`$GITHUB_WORKSPACE` so a relative `dist` resolves correctly there — but making
all references absolute matches the proven-working reference and removes the
ambiguity.)

### Verification

This path only executes in `release.yml` (workflow_run after `main`, on master
pushes and tags), so it cannot be exercised by the PR/unit CI. Verification: the
nightly `release` run on master after merge must reach the GitHub development
release with `dist/` populated; the eventual `v26.6.1` tag run must download the
verified artifact and publish to PyPI.

---

## Part B — Feature: per-release codenames (e.g. "Crossbar.io 26.6.1 — Resistance Is Futile")

### Summary

Give each Crossbar.io CalVer release a maintainer-chosen **codename**, surfaced
in the product title / CLI banner. For example the `26.6` series:

```
Crossbar.io 26.6.1 — Resistance Is Futile
```

with codename **"Resistance Is Futile"**. Codenames are assigned per CalVer
`year.month` series (all `26.6.x` micro releases share one codename).

`crossbar version` and `crossbar keys` output is otherwise unchanged; only the
product/title line gains the codename.

### Motivation

- Memorable, human-friendly release identity alongside the precise CalVer.
- Nice for release announcements, docs, and the startup banner.

### Proposed design

**1. New source file `src/crossbar/_codename.py`**

A sibling of `src/crossbar/_version.py` (same license header, **pure data +
stdlib only** — it is imported at `Personality` class-definition time, so it must
not pull in heavy or internal modules). It holds:

- `_CODENAMES: dict[tuple[int, int], str]` — map of `(year, month) -> codename`,
  e.g. `{(26, 6): "Resistance Is Futile"}`.
- `get_codename(version: str | None = None) -> str | None` — look up the codename
  for a CalVer string (defaults to `crossbar.__version__`). It extracts
  `year, month` the same way the release-key lookup already does
  (`crossbar.common.key._read_release_key`: `version.split(".")[0:2]`),
  `int()`-casting each so `"06"`/`"6"` both work.

**Graceful fallback (mandatory):** if `(year, month)` is not in the map (a dev
build, or a release whose codename hasn't been assigned yet), `get_codename()`
returns `None` and callers omit the codename. It must **never raise** —
`crossbar version` and node startup must keep working (same rule as the
per-release signing key, also derived from `year.month`).

**2. Compose the title once, reuse it**

Add a small helper (in `_codename.py` or `personality.py`) returning the full
product title:

```
release_title()  ->  "Crossbar.io v26.6.1 — Resistance Is Futile [20260620-abc1234]"   # codename present
                 ->  "Crossbar.io v26.6.1 [20260620-abc1234]"                            # codename absent
```

Single source of truth, used by:

- `src/crossbar/personality.py` `_BANNER` — the line currently rendered as
  `{title} v{version} [{build}]` (personality.py:211, formatted at line 236).
- `src/crossbar/node/controller.py:200` — the management-API node title
  (`"{} {}".format(personality.TITLE, crossbar.__version__)`), for consistency.
  (optional but nice)

**3. CLI banner change**

`crossbar version` banner line changes from:

```
    :::::::::::::::::   Crossbar.io v26.4.1.dev1 [19000101-0000000]
```

to (DECIDED: keep the build hash, codename placed before it):

```
    :::::::::::::::::   Crossbar.io v26.6.1 — Resistance Is Futile [20260620-abc1234]
```

When no codename is mapped, the `—` segment is omitted and the line is unchanged
from today's `{title} v{version} [{build}]`.

**4. Tests** — `test_codename.py` (trial, CPython + PyPy):

- `get_codename("26.6.1") == "Resistance Is Futile"`
- dev-suffix handling: `get_codename("26.6.1.dev1") == "Resistance Is Futile"`
- unknown series: `get_codename("19.1.1") is None`
- the rendered banner / `release_title()` contains the codename for a mapped
  version, and omits the `—` separator when absent.

**5. Missing-codename release gate** (DECIDED: yes)

In addition to the always-on soft fallback, add a CI/test gate that **fails when
a non-dev release** (`__version__` without a `.devN` suffix) has **no** codename
in `_codename.py`. This makes a codename mandatory for stable releases — a
stable release can't be tagged unnamed — mirroring how the per-release signing
key is required. Dev builds (`.devN`) are exempt and just fall back to no
codename.

**6. Release process hook**

Add a reminder to the `prep-release` justfile recipe hints (next to the "create a
new release key" step) to ensure a codename for the new `year.month` exists in
`_codename.py` before tagging.

### Decisions (resolved)

1. **Build hash:** KEEP it in the banner, codename placed before the `[build]`
   segment.
2. **Missing-codename policy:** soft fallback in code **plus** a CI/test gate
   failing non-dev releases without a codename.
3. **Codename for 26.6:** `"Resistance Is Futile"`. Future series add their
   codename to `_codename.py` at release-prep time.

### Scope / non-goals

- No change to `__version__` / CalVer semantics; the codename is purely additive
  metadata.
- No change to `crossbar keys` output.

---

_Note: This work was completed with AI assistance (Claude Code). Brought up and reviewed by me (human). _


## Checklist

- [x] I have searched existing issues to avoid duplicates
- [x] I have described the problem clearly
- [x] I have provided use cases
- [x] I have considered alternatives
- [x ] I have assessed impact and breaking changes

--- a/.audit/oberstet_fix_2227.md
+++ b/.audit/oberstet_fix_2227.md
@@ -1,0 +1,8 @@
+- [ ] I did **not** use any AI-assistance tools to help create this pull request.
+- [x] I **did** use AI-assistance tools to *help* create this pull request.
+- [x] I have read, understood and followed the projects' [AI Policy](https://github.com/crossbario/autobahn-python/blob/main/AI_POLICY.md) when creating code, documentation etc. for this pull request.
+
+Submitted by: @oberstet
+Date: 2026-06-21
+Related issue(s): #2227
+Branch: oberstet:fix_2227

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -132,18 +132,18 @@ jobs:
         uses: wamp-proto/wamp-cicd/actions/download-artifact-verified@main
         with:
           name: ${{ needs.check-main-workflow.outputs.artifact_dist }}
-          path: dist
+          path: ${{ github.workspace }}/dist
           run-id: ${{ needs.check-main-workflow.outputs.main_run_id }}
 
       - name: List downloaded packages
         run: |
           echo "Downloaded packages:"
-          ls -lh dist/
+          ls -lh ${{ github.workspace }}/dist/
 
       - name: Validate and prepare release fileset for GitHub
         uses: wamp-proto/wamp-cicd/actions/check-release-fileset@main
         with:
-          distdir: dist
+          distdir: ${{ github.workspace }}/dist
           mode: strict
           keep-metadata: true  # Keep CHECKSUMS for user verification
           targets: |
@@ -216,18 +216,18 @@ jobs:
         uses: wamp-proto/wamp-cicd/actions/download-artifact-verified@main
         with:
           name: ${{ needs.check-main-workflow.outputs.artifact_dist }}
-          path: dist
+          path: ${{ github.workspace }}/dist
           run-id: ${{ needs.check-main-workflow.outputs.main_run_id }}
 
       - name: List downloaded packages
         run: |
           echo "Downloaded packages:"
-          ls -lh dist/
+          ls -lh ${{ github.workspace }}/dist/
 
       - name: Validate and prepare release fileset for GitHub
         uses: wamp-proto/wamp-cicd/actions/check-release-fileset@main
         with:
-          distdir: dist
+          distdir: ${{ github.workspace }}/dist
           mode: strict
           keep-metadata: true  # Keep CHECKSUMS for user verification
           targets: |
@@ -288,7 +288,7 @@ jobs:
       - name: Validate and clean release fileset for PyPI
         uses: wamp-proto/wamp-cicd/actions/check-release-fileset@main
         with:
-          distdir: dist
+          distdir: ${{ github.workspace }}/dist
           mode: strict
           # keep-metadata: false (default - removes CHECKSUMS, build-info.txt etc for PyPI)
           targets: |

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -8,6 +8,8 @@ Changelog
 26.6.1
 ------
 
+* new: per-release codenames (e.g. Crossbar.io 26.6.1 — "Resistance Is Futile"), shown in the ``crossbar version`` banner (`#2227 <https://github.com/crossbario/crossbar/issues/2227>`_)
+* fix: use an absolute path for the verified artifact download in ``release.yml`` so the development/production release jobs find ``dist/`` (`#2227 <https://github.com/crossbario/crossbar/issues/2227>`_)
 * new: coordinated WAMP **26.6.x release train** — pin the coordinated dependency floor: ``autobahn>=26.6.2`` (cryptosign import fix), ``txaio>=26.6.1``, ``zlmdb>=26.6.1``, ``cfxdb>=26.6.1``, ``xbr>=26.6.1`` (`#2224 <https://github.com/crossbario/crossbar/issues/2224>`_)
 * new: track the application ``uv.lock`` for reproducible builds (`#2224 <https://github.com/crossbario/crossbar/issues/2224>`_)
 * new: Crossbar.io 26.6 release key (``crossbar-26-6.pub``)

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -9,6 +9,7 @@ Changelog
 ------
 
 * new: per-release codenames (e.g. Crossbar.io 26.6.1 — "Resistance Is Futile"), shown in the ``crossbar version`` banner (`#2227 <https://github.com/crossbario/crossbar/issues/2227>`_)
+* fix: filter the benign NumPy 1.x/2.x ABI notice printed by bjdata's C extension (autobahn's UBJSON backend) under NumPy 2.x, so it no longer pollutes ``crossbar`` CLI output (short-term shim until bjdata ships a NumPy 2.x build) (`#2227 <https://github.com/crossbario/crossbar/issues/2227>`_)
 * fix: use an absolute path for the verified artifact download in ``release.yml`` so the development/production release jobs find ``dist/`` (`#2227 <https://github.com/crossbario/crossbar/issues/2227>`_)
 * new: coordinated WAMP **26.6.x release train** — pin the coordinated dependency floor: ``autobahn>=26.6.2`` (cryptosign import fix), ``txaio>=26.6.1``, ``zlmdb>=26.6.1``, ``cfxdb>=26.6.1``, ``xbr>=26.6.1`` (`#2224 <https://github.com/crossbario/crossbar/issues/2224>`_)
 * new: track the application ``uv.lock`` for reproducible builds (`#2224 <https://github.com/crossbario/crossbar/issues/2224>`_)

--- a/justfile
+++ b/justfile
@@ -326,8 +326,10 @@ prep-release:
     @echo '   5. Bump to new dev version:  just bump-next <NEXT-VERSION>.dev1'
     @echo '   6. Create a new rel key:     signify-openbsd -G -c "Crossbar.io 26.6" -p src/crossbar/common/keys/crossbar-26-6.pub -s crossbar-26-6.sec'
     @echo '   7. Test rel key:             crossbar version'
+    @echo '   8. Add release codename:     ensure src/crossbar/_codename.py has a codename for this year.month series'
     @echo ''
     @echo 'For rel key: this must be done (incl. *.sec key handling) by maintainer!'
+    @echo 'For codename: a stable release without a codename is rejected by the release gate (test_codename.py).'
     @echo ''
 
 # Auto-bump to next dev version based on current date (CalVer: YY.M.1.dev1)

--- a/src/crossbar/__init__.py
+++ b/src/crossbar/__init__.py
@@ -138,6 +138,41 @@ Command help: {executable} <command> --help
 """
 
 
+def _import_serializers_quietly():
+    """
+    Pre-import ``autobahn.wamp.serializer``, filtering the benign NumPy 1.x/2.x
+    ABI notice that bjdata's C extension prints to stderr.
+
+    Autobahn's optional UBJSON serializer is backed by ``bjdata`` (CPython-only).
+    bjdata ships sdist-only and builds its ``_bjdata`` C extension against
+    NumPy 1.x (via the deprecated ``oldest-supported-numpy``), so importing it
+    under NumPy 2.x prints a multi-line "compiled using NumPy 1.x" notice to
+    stderr. It is harmless: only bjdata's (unused) NumPy-array path is affected;
+    UBJSON (de)serialization of WAMP payloads works. Importing the serializer
+    module once, early and quietly, keeps the notice out of the ``crossbar`` CLI
+    output - later imports hit the module cache and stay silent.
+
+    This is NOT caught by ``warnings.filterwarnings`` (NumPy writes directly to
+    stderr), hence the stderr redirect.
+
+    TODO: remove this shim once bjdata builds against NumPy 2.x / ships wheels.
+    Upstream: https://github.com/NeuroJSON/pybj
+    """
+    import contextlib
+    import io
+
+    captured = io.StringIO()
+    try:
+        with contextlib.redirect_stderr(captured):
+            import autobahn.wamp.serializer  # noqa: F401
+    finally:
+        # re-emit anything that is NOT the known benign NumPy 1.x/2.x notice,
+        # so we never silently swallow unexpected output from this import
+        text = captured.getvalue()
+        if text and "compiled using NumPy 1.x" not in text:
+            sys.stderr.write(text)
+
+
 def run():
     """
      CLI entry point into crossbar.
@@ -300,6 +335,10 @@ def run():
         reactor = install_reactor(
             explicit_reactor=os.environ.get("CROSSBAR_REACTOR", None), verbose=False, require_optimal_reactor=False
         )
+
+        # quietly pre-import the WAMP serializers before the personality import
+        # chain pulls them in, to filter bjdata's benign NumPy 1.x/2.x notice
+        _import_serializers_quietly()
 
         # get chosen personality class
         if command == "standalone":

--- a/src/crossbar/_codename.py
+++ b/src/crossbar/_codename.py
@@ -1,0 +1,66 @@
+#####################################################################################
+#
+#  Copyright (c) typedef int GmbH
+#  SPDX-License-Identifier: EUPL-1.2
+#
+#####################################################################################
+
+"""
+Crossbar.io release codenames.
+
+Each Crossbar.io CalVer release series (``year.month``) carries a
+maintainer-chosen codename, e.g. ``26.6.x`` is "Resistance Is Futile". The
+codename is purely additive metadata surfaced in the product title / CLI banner;
+it does not affect versioning.
+
+A new codename is assigned per ``year.month`` series at release-prep time (see
+the ``prep-release`` recipe in the justfile), alongside the version bump and the
+new release signing key.
+"""
+
+from typing import Dict, Optional, Tuple
+
+# Map of (year, month) CalVer series -> release codename.
+#
+# Add a new entry for every new year.month release series. A stable (non-dev)
+# release without a codename is rejected by the release gate (see
+# crossbar/test/test_codename.py).
+_CODENAMES: Dict[Tuple[int, int], str] = {
+    (26, 6): "Resistance Is Futile",
+}
+
+
+def get_codename(version: Optional[str] = None) -> Optional[str]:
+    """
+    Look up the release codename for a Crossbar.io CalVer version string.
+
+    :param version: CalVer version string like ``"26.6.1"`` or
+        ``"26.6.1.dev1"``. Defaults to the running ``crossbar.__version__``.
+    :returns: The codename for the version's ``(year, month)`` series, or
+        ``None`` if no codename is assigned (e.g. a development build, or a
+        release that has not been named yet). Never raises.
+    """
+    if version is None:
+        import crossbar
+
+        version = crossbar.__version__
+
+    try:
+        parts = version.split(".")
+        year = int(parts[0])
+        month = int(parts[1])
+    except (AttributeError, IndexError, ValueError):
+        return None
+
+    return _CODENAMES.get((year, month))
+
+
+def get_codename_suffix(version: Optional[str] = None) -> str:
+    """
+    Return the codename as a title/banner suffix, e.g. ``" — Resistance Is
+    Futile"``, or the empty string when no codename is assigned.
+
+    :param version: see :func:`get_codename`.
+    """
+    codename = get_codename(version)
+    return " — {}".format(codename) if codename else ""

--- a/src/crossbar/node/controller.py
+++ b/src/crossbar/node/controller.py
@@ -28,6 +28,7 @@ from twisted.python.runtime import platform
 from txaio import get_global_log_level, make_logger
 
 import crossbar
+from crossbar._codename import get_codename_suffix
 from crossbar._util import class_name, hl, hlid, hltype, hlval, term_print
 from crossbar.common.checkconfig import (
     NODE_SHUTDOWN_ON_LAST_WORKER_EXIT,
@@ -197,7 +198,7 @@ class NodeController(NativeProcess):
             workers_by_type[worker.TYPE] += 1
 
         return {
-            "title": "{} {}".format(self.personality.TITLE, crossbar.__version__),
+            "title": "{} {}{}".format(self.personality.TITLE, crossbar.__version__, get_codename_suffix()),
             "started": self._started,
             "controller_pid": self._pid,
             "running_workers": len(self._workers),

--- a/src/crossbar/personality.py
+++ b/src/crossbar/personality.py
@@ -14,6 +14,7 @@ import txaio
 txaio.use_twisted()
 
 import crossbar
+from crossbar._codename import get_codename_suffix
 from crossbar.common import checkconfig
 from crossbar.interfaces import IInventory, IRealmStore
 from crossbar.node import node
@@ -208,7 +209,7 @@ _BANNER = r"""
     :::::::   :::::::  / /__/ __/ _ \(_-<(_-</ _ \/ _ `/ __/ / / _ \
     :::::   :   :::::  \___/_/  \___/___/___/_.__/\_,_/_/ (_)_/\___/
           :::::
-    :::::::::::::::::   {title} v{version} [{build}]
+    :::::::::::::::::   {title} v{version}{codename} [{build}]
 
     Copyright (c) 2013-{year} typedef int GmbH. Licensed under EUPLv1.2.
 """
@@ -234,7 +235,11 @@ class Personality(object):
     DESC = _DESC
 
     BANNER = _BANNER.format(
-        title=_TITLE, version=crossbar.__version__, build=crossbar.__build__, year=time.strftime("%Y")
+        title=_TITLE,
+        version=crossbar.__version__,
+        codename=get_codename_suffix(crossbar.__version__),
+        build=crossbar.__build__,
+        year=time.strftime("%Y"),
     )
 
     LICENSE = ("crossbar", "LICENSE")

--- a/src/crossbar/test/test_codename.py
+++ b/src/crossbar/test/test_codename.py
@@ -1,0 +1,64 @@
+#####################################################################################
+#
+#  Copyright (c) typedef int GmbH
+#  SPDX-License-Identifier: EUPL-1.2
+#
+#####################################################################################
+
+from twisted.trial import unittest
+
+import crossbar
+from crossbar._codename import get_codename, get_codename_suffix
+
+
+class TestCodename(unittest.TestCase):
+    """
+    Tests for the per-release codename lookup. See issue #2227.
+    """
+
+    def test_known_series(self):
+        self.assertEqual(get_codename("26.6.1"), "Resistance Is Futile")
+
+    def test_dev_suffix(self):
+        # a .devN suffix on a known year.month series still resolves
+        self.assertEqual(get_codename("26.6.1.dev1"), "Resistance Is Futile")
+
+    def test_unknown_series_returns_none(self):
+        self.assertIsNone(get_codename("19.1.1"))
+
+    def test_malformed_version_returns_none(self):
+        # must never raise, just return None
+        self.assertIsNone(get_codename("garbage"))
+        self.assertIsNone(get_codename("26"))
+        self.assertIsNone(get_codename(""))
+
+    def test_default_version(self):
+        # called without an argument, uses crossbar.__version__ and must not raise
+        self.assertEqual(get_codename(), get_codename(crossbar.__version__))
+
+    def test_suffix_present(self):
+        self.assertEqual(get_codename_suffix("26.6.1"), " — Resistance Is Futile")
+
+    def test_suffix_absent(self):
+        self.assertEqual(get_codename_suffix("19.1.1"), "")
+
+    def test_banner_contains_codename(self):
+        # the running version is a 26.6.x release, which has a codename, so the
+        # startup banner must include it
+        from crossbar.personality import Personality
+
+        codename = get_codename(crossbar.__version__)
+        if codename is None:
+            raise unittest.SkipTest("running version {} has no codename".format(crossbar.__version__))
+        self.assertIn(codename, Personality.BANNER)
+
+    def test_release_gate_stable_has_codename(self):
+        # release gate: a stable (non-dev) release MUST have a codename assigned
+        # in crossbar/_codename.py. Development builds (.devN) are exempt.
+        version = crossbar.__version__
+        if ".dev" in version:
+            raise unittest.SkipTest("development build {}; codename optional".format(version))
+        self.assertIsNotNone(
+            get_codename(version),
+            "stable release {} has no codename - add one to crossbar/_codename.py".format(version),
+        )

